node/src/components/linear_chain_sync/operations.rs

```diff
@@ -0,0 +1,288 @@
+
+    // Get the most recent header which has the same version as ours
+    // We keep fetching by height until none of our peers have a block at that height
+    let mut most_recent_block_header = *trusted_header;
+    loop {
+        // If we encounter a block header of a version which is newer than ours we return an error
+        if most_recent_block_header.protocol_version() > chainspec.protocol_config.version {
+            return Err(
+                LinearChainSyncError::RetrievedBlockHeaderFromFutureVersion {
+                    current_version: chainspec.protocol_config.version,
+                    block_header_with_future_version: Box::new(most_recent_block_header),
+                },
+            );
+        }
+        let maybe_fetched_block = fetch_block_header_by_height(
+            effect_builder,
+            most_recent_block_header.height() + 1,
+            &trusted_validator_weights,
+        )
+        .await?;
+        match maybe_fetched_block {
+            Some(more_recent_block_header_with_metadata) => {
+                most_recent_block_header = more_recent_block_header_with_metadata.block_header;
+                // If the new block is a switch block, update the validator weights
+                if let Some(new_trusted_validator_weights) =
+                    most_recent_block_header.next_era_validator_weights()
+                {
+                    trusted_validator_weights = new_trusted_validator_weights.clone();
+                }
+            }
+            // If we could not fetch, we can stop if the most recent has our protocol version
+            None if most_recent_block_header.protocol_version()
+                == chainspec.protocol_config.version =>
+            {
+                break
+            }
+            // Otherwise keep trying to fetch until we get a block with our version
+            _ => {
+                tokio::time::sleep(TIMEOUT_DURATION).await;
+            }
+        }
+    }
+
+    // Synchronize the trie store for the most recent block header.
+
+    // Corner case: if an emergency restart happened recently, it is necessary to synchronize the
+    // state for the block right after the emergency restart.  This is needed for the EraSupervisor.
+
+    // The era supervisor needs validator information from previous eras it may potentially slash.
+    // The number of previous eras is determined by a *delay* in which consensus participants become
+    // bonded validators or unbond.
+    let delay = era_supervisor::bonded_eras(&(&chainspec).into());
+    // The era supervisor may slash up to 3*delay + 1 eras back
+    let historical_eras_needed = delay.saturating_mul(3).saturating_add(1);
+
+    match maybe_last_emergency_restart_era_id {
+        Some(last_emergency_restart_era_id)
+            if last_emergency_restart_era_id
+                > most_recent_block_header
+                    .era_id()
+                    .saturating_sub(historical_eras_needed) =>
+        {
```

Comment:
@Fraser999 @fizyk20 I'm not sure if the logic below does the right thing in this corner case